lib/src/util/bolt_helpers.cc

```diff
@@ -0,0 +1,195 @@
+    // the file is moved to destination with boost::filesystem::rename.
+    //
+    // If the destination param is empty: the function assumes the file should
+    // remain in the cache_dir and will be renamed to cache_dir/filename
+    fs::path downloadFileFromMaster(const std::vector<std::string>& master_uris,
+                                uint32_t connect_timeout,
+                                uint32_t timeout,
+                                lth_curl::client& client,
+                                const fs::path& cache_dir,
+                                const fs::path& destination,
+                                const lth_jc::JsonContainer& file) {
+        auto filename = fs::path(file.get<std::string>("filename")).filename();
+        auto sha256 = file.get<std::string>("sha256");
+        // if destination is empty assume the file should just stay in the cache_dir
+        auto filepath = destination;
+        if (filepath.empty()) {
+            filepath = cache_dir / filename;
+        }
+
+        if (fs::exists(filepath) && sha256 == calculateSha256(filepath.string())) {
+            fs::permissions(filepath, NIX_TASK_FILE_PERMS);
+            return filepath;
+        }
+
+        if (master_uris.empty()) {
+            throw Module::ProcessingError(lth_loc::format("Cannot download task. No master-uris were provided"));
+        }
+
+        auto tempname = cache_dir / fs::unique_path("temp_task_%%%%-%%%%-%%%%-%%%%");
+        // Note that the provided "file_path" argument is a temporary file, call it "tempA".
+        // Leatherman.curl during the download method will create another temporary file,
+        // call it "tempB", to save the downloaded file's contents in chunks before
+        // renaming it to "tempA." The rationale behind this solution is that:
+        //    (1) After download, we still need to check "tempA" to ensure that its sha matches
+        //    the provided sha. So the downloaded task is not quite a "valid" file after this
+        //    method is called; it's still temporary.
+        //
+        //    (2) It somewhat simplifies error handling if multiple threads try to download
+        //    the same file.
+        auto download_result = downloadFileWithCurl(master_uris, connect_timeout, timeout, client, tempname, file.get<lth_jc::JsonContainer>("uri"));
+        if (!std::get<0>(download_result)) {
+            throw Module::ProcessingError(lth_loc::format(
+                "Downloading the task file {1} failed after trying all the available master-uris. Most recent error message: {2}",
+                file.get<std::string>("filename"),
+                std::get<1>(download_result)));
+        }
+
+        if (sha256 != calculateSha256(tempname.string())) {
+            fs::remove(tempname);
+            throw Module::ProcessingError(lth_loc::format("The downloaded {1}'s sha differs from the provided sha", filename));
```

Comment:
Does `sha` need to be `SHA` in user visible messaging? Also, try reading this message out loud with an arbitrary filepath in place of the argument and see if the phrasing feels awkward to you. (Yes, I see that this is code that you moved, but anyways...)

Reply:
Oh, yeah, it should read:

`The downloaded file {1} has a SHA that does not match the provided SHA`